index.rst

```diff
@@ -1,42 +1,198 @@
 
+    :implied by: :extension:`TypeOperators`, :extension:`TypeFamilies`
     :since: 7.6.1
 
-    Enable use of explicit namespaces in module export lists.
+    Enable use of explicit namespaces in module import and export lists.
 
-In an import or export list, such as ::
+Background
+~~~~~~~~~~
 
-      module M( f, (++) ) where ...
-        import N( f, (++) )
-        ...
+Haskell has two namespaces:
 
-the entities ``f`` and ``(++)`` are *values*. However, with type
-operators it becomes possible to declare
-``(++)`` as a *type constructor*. In that case, how would you export or
-import it?
+* the *type namespace*, including the names of type constructors, type
+  synonyms, type families and type classes;
 
-The :extension:`ExplicitNamespaces` extension allows you to prefix the name of
-a type constructor in an import or export list with "``type``" to
-disambiguate this case, thus: ::
+* the *data namespace*, including the names of term-level values, functions,
+  data constructors and pattern synonyms.
 
-      module M( f, type (++) ) where ...
-        import N( f, type (++) )
-        ...
-      module N( f, type (++) ) where
-        data family a ++ b = L a | R b
+This separation allows us to define data constructors and type
+constructors whose names coincide:
 
-The extension :extension:`ExplicitNamespaces` is implied by
-:extension:`TypeOperators` and (for some reason) by :extension:`TypeFamilies`.
+.. code:: haskell
 
-In addition, with :extension:`PatternSynonyms` you can prefix the name of a
-data constructor in an import or export list with the keyword
-``pattern``, to allow the import or export of a data constructor without
-its parent type constructor.
+   data T = T
+
+(Similarly, with the ``TypeOperators`` extension it is possible to define
+term-level operators and type families whose names coincide.)
+
+At use sites, GHC infers which ``T`` is referred to from the context:
+
+.. code:: haskell
+
+   t :: T  -- type-level T
+   t = T   -- term-level T
+
+The use of identical names for type-level and term-level entities is called
+*punning*.  Haskell makes heavy use of punning in its built-in syntax and common
+types.
+
+However, there are various contexts in which an occurrence of a name may refer
+either to the type namespace or the data namespace, and it is not always clear
+which is meant. In particular, the following may mention both type-level and
+term-level entities:
+
+- Import and export lists
+
+- Fixity declarations
+
+- ``WARNING``, ``DEPRECATED`` and ``ANN`` pragmas
+
+- Types, when using the ``DataKinds`` extension to reference a data constructor
+  at the type level
+
+- Template Haskell name quotes
+
+In certain situations, which appear to be geining prevalence in modern Haskell, precison is needed
+in specifying exactly which names are exported or imported from a module. This extension aims to
+provide precise control in the presence of puns where the same name is present in the type and
+data namespace.
```

Comment:
It's good to motivate the problem but this could be cut down a bit I think, the full list is less relevant in the docs than in the proposal.

Reply:
Indeed, that should probably be condensed.